src/pq_flash_index.cpp

```diff
@@ -1167,27 +1226,29 @@ void PQFlashIndex<T, LabelT>::cached_beam_search(const T *query1, const uint64_t
     std::vector<Neighbor> &full_retset = query_scratch->full_retset;
 
     uint32_t best_medoid = 0;
+    uint32_t best_shard = 0;
     float best_dist = (std::numeric_limits<float>::max)();
-    if (!use_filter)
+    for (uint64_t cur_m = 0; cur_m < num_medoids; cur_m++)
     {
-        for (uint64_t cur_m = 0; cur_m < num_medoids; cur_m++)
+        float cur_expanded_dist =
+            dist_cmp_float->compare(query_float, centroid_data + aligned_dim * cur_m, (uint32_t)aligned_dim);
+        if (cur_expanded_dist < best_dist)
         {
-            float cur_expanded_dist =
-                dist_cmp_float->compare(query_float, centroid_data + aligned_dim * cur_m, (uint32_t)aligned_dim);
-            if (cur_expanded_dist < best_dist)
-            {
-                best_medoid = medoids[cur_m];
-                best_dist = cur_expanded_dist;
-            }
+            best_medoid = medoids[cur_m];
+            best_dist = cur_expanded_dist;
+            best_shard = cur_m;
         }
     }
-    else if (_filter_to_medoid_id.find(filter_label) != _filter_to_medoid_id.end())
-    {
-        best_medoid = _filter_to_medoid_id[filter_label];
-    }
-    else
+    if (use_filter)
     {
-        throw ANNException("Cannot find medoid for specified filter.", -1, __FUNCSIG__, __FILE__, __LINE__);
+        if (_filter_to_medoid_ids.find(filter_label) != _filter_to_medoid_ids.end())
+        {
+            best_medoid = _filter_to_medoid_ids[filter_label][best_shard];
```

Comment:
Can we guarantee that each shard contains all kinds of labels? if no, then here may have issue

Reply:
especially, can we guarantee that the best_shard contains this label "filter_label"

Reply:
good point, indeed this issue can arise. Let me think of a fix for this.